### PR TITLE
security: bound NSA paged-indexer cache addresses

### LINKED ISSUE #160: security: make NSA paged-indexer page validation mandatory and complete
## Summary

The NSA paged indexer skips CUDA page-table value validation by default; its opt-in validation still omits the physical upper page bound.

**Audited revision:** `954fd0174c49502b62547a01f09c404029e6035a`

## Threat model / reachability

A caller or cache manager controls `real_page_table` / `cache_seqlens`.

## Evidence

- [`b12x/attention/nsa_indexer/paged.py:386-431`](https://github.com/local-inference-lab/b12x/blob/954fd0174c49502b62547a01f09c404029e6035a/b12x/attention/nsa_indexer/paged.py#L386-L431) returns early unless `B12X_VALIDATE_PAGED_INDEXER_CUDA_VALUES=1` and never compares IDs with cache pages.
- `b12x/attention/nsa_indexer/fused_indexer.py:1519-1533` computes `Int64(pid) * k_quant_page_stride`.
- `b12x/attention/nsa_indexer/fused_indexer.py:1644-1651` accepts every nonnegative page ID.

## Impact

Out-of-bounds quantized K/scale reads, GPU fault, denial of service, or adjacent device-data exposure via scores/top-k.

## Recommended fix

- Make validation a binding invariant, not an env opt-in.
- Validate IDs against value and scale cache extents.
- Keep a kernel upper-bound predicate and add high-pid tests.

## Verification notes

Source review at the audited revision; runtime exploitation was not attempted unless stated below.

### PR DESCRIPTION
**Pull request by malaiwah**
## Problem

NSA paged-indexer kernels accepted non-negative page-table IDs without proving the physical page existed in the packed K/value and scale cache. CUDA graph replay keeps page-table values device-live, so host-only validation could not make the launch safe.

## Fix

- bound every direct fused, legacy scalar/tiled, and streamed scorer page ID against both live cache tensors before address formation
- pass runtime cache capacity into the packed-contiguous gather and mask invalid pages before loading
- pass runtime cache capacity into final logical-to-physical top-k emission and emit `-1` for invalid physical pages
- preserve graph replay, runtime row-stride reuse, compile-cache identity, and valid-page behavior
- add invalid-page graph-replay coverage for fused CTA policies, legacy/stream tiled paths, packed-contiguous gather, and physical-slot mapping

Closes #160

## Verification

- SM121: 12 targeted GPU tests passed at final head `f7d51b7`, including graph replay and the existing live high-page-ID Int64-offset regression
- focused `py_compile`, Ruff (excluding pre-existing diagnostics), and `git diff --check` passed
- independent peer review: APPROVE; all scoped sinks are bounded by live value/scale capacity before access or physical emission

## Summary by CodeRabbit

- **Bug Fixes**
  - Invalid or out-of-range paged cache IDs are now safely rejected across attention indexing and scoring paths.
  - Prevented invalid cache entries from producing out-of-bounds accesses, incorrect scores, or physical slots.
  - Improved handling of invalid entries during tiled top-k processing and CUDA graph replay.

- **Tests**
  - Added regression and integration coverage for invalid page IDs, cache boundaries, physical-slot output, and streaming scorer modes.

### CODE REVIEW

**Comment by malaiwah** [comment]
@coderabbitai review

**Comment by malaiwah** [comment]
## Aggregate integration qualification evidence

This PR's patch was included in the combined security qualification branch [`malaiwah/sparkinfer:security/integrated-qualification`](https://github.com/malaiwah/sparkinfer/tree/security/integrated-qualification).

- Qualified integration commit: `01d74106b9efbff9bce1ecd66893b4cbadc0c919`
- Integration baseline/merge-base: `195e26c5b67eb9c162f00fd8907d06e0fe27c569`
- Environment: PyTorch `2.13.0+cu130`, CUDA 13.0, CUTLASS DSL 4.6.0
- Hardware: 2× NVIDIA RTX PRO 6000 Blackwell Server Edition, 97,887 MiB each
- Benchmark/test GPU UUID: `GPU-554f7aeb-578c-187f-4f29-af06f2170507`

The combined changed-path GPU run exercised compiler-cache, attention, page/index validation, stream handling, PCIe IPC/DCP, GEMM, MoE routing, FP6 artifact/export/dequantization, profiler, packaging, and release paths:

```text
1011 passed, 4 skipped, 1 warning in 125.69s
```

The artifact/profiler/release subset independently completed:

```text
403 passed, 2 skipped in 38.48s
```

With CUDA hidden, the CPU-compatible security set completed:

```text
525 passed, 59 skipped in 39.75s
```

Correctness gates passed for the benchmarked varlen attention paths: baseline and integrated arms produced identical checksums and maximum magnitude, finite tensors, and nonzero output. Balanced ABBA timings on the GPU above measured:

| Case | Baseline run means | Integrated run means | Direction |
|---|---:|---:|---:|
| Varlen `(128,256,64,512)`, QH=4, KVH=2, D=64 | 20.888, 21.024 µs | 66.070, 64.094 µs | integrated **3.106× slower**, +44.13 µs |
| Varlen `4×512`, QH=8, KVH=2, D=128 | 24.787, 24.813 µs | 67.469, 63.840 µs | integrated **2.647× slower**, +40.85 µs |

The fixed eager-launch cost comes from graph-safe validation/sanitization kernels. Standalone MoE route sanitization measured 7.387/7.384/7.322 µs for 8/512/4096 routes. Invalid-row scrub measured 6.967/6.909/7.031 µs for 8/64/512 rows at K=4096.

Memory deltas observed in qualification:

- short varlen scratch: 506,880 → 508,932 bytes, +2,052 bytes (+0.405%)
- prefill varlen scratch: 4,259,840 → 4,261,892 bytes, +2,052 bytes (+0.048%)
- MoE route sanitizer: 8 bytes per route; 32 KiB at the default 4096-route capacity

Scope: this is evidence for the exact aggregate integration commit, not a claim that a later PR-head update was independently rerun. The full repository test tree is not CPU-compatible, and no end-to-end vLLM/SGLang model-serving run was performed. The temporary GPU runner was destroyed after qualification.

**Review by lukealonso — CHANGES_REQUESTED** [review]
The address hardening adds hot-kernel branches and 64-bit arithmetic across NSA index paths and also changes `preinitialize_tile_logits` from false to true. The bounds checks may be necessary, but the initialization-policy change can add work independently and the combined register, occupancy, and launch impact is not characterized.

Please rebase on the merged NSA security work in #184, separate or justify the initialization-policy change, and provide correctness plus SASS/resource and target-GPU decode/prefill measurements, including big-page-id coverage beyond the 32-bit offset boundary and CUDA-graph replay.

**Comment by lukealonso** [comment]
Closing rather than merging. The implementation changes NSA device-kernel addressing and replay-time predicates across multiple scorer and gather paths. That directly violates the zero-kernel-code-change constraint. The issue remains open; it cannot be represented as a host-only security fix while mutable graph-replay page IDs remain in scope.
